### ISSUE TITLE
can 3060 12 gb run this project？

### ISSUE THREAD

**Issue by KCKCC** [thu-ml/TurboDiffusion#2]
the result speed is awesome, my comfy can run wan2.2 
is it possible runnable for 3060？

**Comment by xiaoxuexiaoniu** [comment]
针对TurboWan2.1-T2V-1.3B-480P-quant.pth模型，经过验证，将源码的VAE解码分块处理采样 + bfloat16处理，是可以实现在单块或多块3060(12gVRAM)成功单次推理，就是视频质量低点，方案可行，等后续有量化版本出来，应该会好点

**Comment by xiaoxuexiaoniu** [comment]
下面是源码修改：

import argparse
import math
import time

import torch
from einops import rearrange, repeat
from tqdm import tqdm

from imaginaire.utils.io import save_image_or_video
from imaginaire.utils import log

from rcm.datasets.utils import VIDEO_RES_SIZE_INFO
from rcm.utils.umt5 import clear_umt5_memory, get_umt5_embedding
from rcm.tokenizers.wan2pt1 import Wan2pt1VAEInterface, Wan2pt1VAEInterface2

from modify_model import tensor_kwargs, create_model

torch._dynamo.config.suppress_errors = True

def parse_arguments() -> argparse.Namespace:
    parser = argparse.ArgumentParser(description="TurboDiffusion inference script for Wan2.1 T2V")
    parser.add_argument("--dit_path", type=str, required=False, help="Custom path to the DiT model checkpoint for distilled models")
    parser.add_argument("--model", choices=["Wan2.1-1.3B", "Wan2.1-14B"], default="Wan2.1-1.3B", help="Model to use")
    parser.add_argument("--num_samples", type=int, default=1, help="Number of samples to generate")
    parser.add_argument("--num_steps", type=int, choices=[1, 2, 3, 4], default=4, help="1~4 for timestep-distilled inference")
    parser.add_argument("--sigma_max", type=float, default=80, help="Initial sigma for rCM")
    parser.add_argument("--vae_path", type=str, default="/home/develop/project/py_project/ComfyUI/Fast-Wan/TurboDiffusion/checkpoints/Wan2.1_VAE.pth", help="Path to the Wan2.1 VAE")
    parser.add_argument("--text_encoder_path", type=str, default="/home/develop/project/py_project/ComfyUI/Fast-Wan/TurboDiffusion/checkpoints/models_t5_umt5-xxl-enc-bf16.pth", help="Path to the umT5 text encoder")
    parser.add_argument("--num_frames", type=int, default=81, help="Number of frames to generate")
    parser.add_argument("--prompt", type=str, help="Text prompt for video generation (required unless --serve)")
    parser.add_argument("--resolution", default="480p", type=str, help="Resolution of the generated output")
    parser.add_argument("--aspect_ratio", default="16:9", type=str, help="Aspect ratio of the generated output (width:height)")
    parser.add_argument("--seed", type=int, default=0, help="Random seed for reproducibility")
    parser.add_argument("--save_path", type=str, default="output/generated_video.mp4", help="Path to save the generated video (include file extension)")
    parser.add_argument("--attention_type", choices=["sla", "sagesla", "original"], default="sagesla", help="Type of attention mechanism to use")
    parser.add_argument("--sla_topk", type=float, default=0.1, help="Top-k ratio for SLA/SageSLA attention")
    parser.add_argument("--quant_linear", action="store_true", help="Whether to replace Linear layers with quantized versions")
    parser.add_argument("--default_norm", action="store_true", help="Whether to replace LayerNorm/RMSNorm layers with faster versions")
    parser.add_argument("--serve", action="store_true", help="Launch interactive TUI server mode (keeps model loaded)")
    return parser.parse_args()

def chunk_decode(tokenizer, latent, device_t="cuda:1", chunk_size=4):
    B, C, T, H, W = latent.shape
    decoded_chunks = []
    with torch.no_grad(), torch.autocast(device_type=device_t,dtype=torch.bfloat16):
         for t_start in range(0, T, chunk_size):
            t_end = min(t_start + chunk_size, T)
            latent_chunk = latent[:,:, t_start:t_end ,:,:]
            chunk = tokenizer.decode(latent_chunk)
            decoded_chunks.append(chunk)
            torch.cuda.empty_cache()
    return torch.cat(decoded_chunks,dim=2)

if __name__ == "__main__":
    args = parse_arguments()
    # Handle serve mode
    if args.serve:
        # Set mode to t2v for the TUI server
        args.mode = "t2v"
        from serve.tui import main as serve_main
        serve_main(args)
        exit(0)

    # Validate prompt is provided for one-shot mode
    if args.prompt is None:
        log.error("--prompt is required (unless using --serve mode)")
        exit(1)

    log.info(f"Computing embedding for prompt: {args.prompt}")
    with torch.no_grad():
        text_emb = get_umt5_embedding(checkpoint_path=args.text_encoder_path, prompts=args.prompt).to(**tensor_kwargs)
    clear_umt5_memory()

    log.info(f"Loading DiT model from {args.dit_path}")
    net = create_model(dit_path=args.dit_path, args=args).cpu()
    torch.cuda.empty_cache()
    log.success("Successfully loaded DiT model.")
    
    tokenizer = Wan2pt1VAEInterface(vae_pth=args.vae_path)

    w, h = VIDEO_RES_SIZE_INFO[args.resolution][args.aspect_ratio]

    log.info(f"Generating with prompt: {args.prompt}")
    condition = {"crossattn_emb": repeat(text_emb.to(**tensor_kwargs), "b l d -> (k b) l d", k=args.num_samples)}

    to_show = []

    state_shape = [
        tokenizer.latent_ch,
        tokenizer.get_latent_num_frames(args.num_frames),
        h // tokenizer.spatial_compression_factor,
        w // tokenizer.spatial_compression_factor,
    ]

    generator = torch.Generator(device=tensor_kwargs["device"])
    generator.manual_seed(args.seed)

    init_noise = torch.randn(
        args.num_samples,
        *state_shape,
        dtype=torch.bfloat16,
        device=tensor_kwargs["device"],
        generator=generator,
    )

    # mid_t = [1.3, 1.0, 0.6][: args.num_steps - 1]
    # For better visual quality
    mid_t = [1.5, 1.4, 1.0][: args.num_steps - 1]

    t_steps = torch.tensor(
        [math.atan(args.sigma_max), *mid_t, 0],
        dtype=torch.bfloat16,
        device=init_noise.device,
    )

    # Convert TrigFlow timesteps to RectifiedFlow
    t_steps = torch.sin(t_steps) / (torch.cos(t_steps) + torch.sin(t_steps))

    # Sampling steps
    x = init_noise.to(torch.bfloat16) * t_steps[0]
    ones = torch.ones(x.size(0), 1, device=x.device, dtype=x.dtype)
    total_steps = t_steps.shape[0] - 1
    net.cuda()
    for i, (t_cur, t_next) in enumerate(tqdm(list(zip(t_steps[:-1], t_steps[1:])), desc="Sampling", total=total_steps)):
        with torch.no_grad():
            v_pred = net(x_B_C_T_H_W=x.to(**tensor_kwargs), timesteps_B_T=(t_cur.float() * ones * 1000).to(**tensor_kwargs), **condition).to(
                torch.bfloat16
            )
            x = (1 - t_next) * (x - t_cur * v_pred) + t_next * torch.randn(
                *x.shape,
                dtype=torch.bfloat16,
                device=tensor_kwargs["device"],
                generator=generator,
            )
    samples = x.bfloat16()
    net.cpu()
    torch.cuda.empty_cache()
    
    # new tokenizer in cuda:1
    # new_tokenizer = Wan2pt1VAEInterface2(vae_pth=args.vae_path)
    # torch.cuda.empty_cache()
    
    with torch.no_grad():
        # new_samples = samples.to("cuda:1")
        # video = chunk_decode(new_tokenizer, new_samples, device_t="cuda:1", chunk_size=1)
        video = chunk_decode(tokenizer, samples, device_t="cuda:0",chunk_size=1)
    to_show.append(video.float().cpu())

    to_show = (1.0 + torch.stack(to_show, dim=0).clamp(-1, 1)) / 2.0
    
    # 获取当前时间戳
    current_timestamp = time.time()
    print(current_timestamp)

    # 获取当前时间（格式为字符串）
    current_time_str = time.strftime('%Y%m%d%H%M%S', time.localtime(current_timestamp))
    print(current_time_str)

    save_image_or_video(rearrange(to_show, "n b c t h w -> c t (n h) (b w)"), "output/generated_video" + current_time_str + ".mp4", fps=16)

> 

>
